File: 2022/day01.py
from itertools import groupby

def read_input_lines():
    with open("day01.txt", 'r') as f:
        return (x.strip() for x in f.readlines())

def get_calories_list():
    lines = read_input_lines()
    calories_groups = (x for k,x in groupby(lines, lambda c: c == '') if not k)
    int_list = (list(map(int, x)) for x in calories_groups)
    return map(sum, int_list)

def part1():
    calories_list = get_calories_list()
    max_calories = max(calories_list)
    print(max_calories)

def part2():
    calories_list = list(sorted(get_calories_list()))
    max_three = sum(calories_list[-3:])
    print(max_three)


if __name__ == "__main__":
    part1()
    part2()

File: 2022/day01.rs
fn read_input() -> Vec<String> {
    std::fs::read_to_string("day01.txt")
        .unwrap()
        .lines()
        .map(String::from)
        .collect()
}

fn get_calories_list() -> Vec<i32> {
    let input = read_input();
    let output = input.into_iter().fold(vec![0], |mut acc, x| {
        if x == "" {
            acc.push(0);
        } else {
            *acc.last_mut().unwrap() += x.parse::<i32>().unwrap();
        }
        acc
    });
    output
}

fn part1() -> i32 {
    let calories_list = get_calories_list();
    calories_list.into_iter().max().unwrap()
}

fn part2() -> i32 {
    let mut calories_list = get_calories_list();
    calories_list.sort();
    calories_list.into_iter().rev().take(3).sum()
}

fn main() {
    println!("{}", part1());
    println!("{}", part2());
}
